lsコマンド新規追加

ItohYuto opened the pull request:
コマンド実行時のカレントディレクトリ内のファイル一覧を表示します。
.ファイルは表示されません。
列幅は最大文字数＋1です。

レビューお願いいたします。

ItohYuto committed:
lsコマンド新規追加

コマンド実行時のカレントディレクトリ内のファイル一覧を表示します。
.ファイルは表示されません。

ItohYuto committed:
配列名を適切な名前に変更。
列幅を最大文字数+1に変更。

ItohYuto committed:
不要なコメント削除

u1tnk commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,49 @@
+  displays = setup_array(files)
+  max_file_name_length = files.map(&:size).max
+  cnt = 0
+  catch :files_end do
```
正直この書き方知りませんでした 🙈 
とはいえ業務で書くときに滅多に見ない書き方だと読みづらくなっちゃうのでruby業務っぽい書き方も覚えて欲しいです。
このロジックだと whileとuntileあたりがそのまま書き替えられそうですが、

そもそも `file.size.times` ではダメでした？

ItohYuto replied:
11行目のcatchに関しては、ループをさせるために実装しているのではなく12行目、13行目のループの途中で抜けるために実装しており、```file.size.times```や```while```、```until```でもう一つ外にループを作ってしまった場合、今の実装から思いつきませんでした。

catch,throwを使用して大域脱出で実装しましたが、今回はループ分を全部抜ける＝メソッドを抜けるなので、```return```でメソッドごと抜けることにしました。

u1tnk replied:
そうですね、  'return' なら問題無いですね。
ちなみに catch,throwでも以下の形式の例外処理は普通に使います。

```
begin
  ...
  raise
rescue
end
```

とはいえループやreturnで書けて読みやすさも大差無いなら例外処理は使わない方が良いですね。
多少ですがパフォーマンスに影響するので。

u1tnk commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,49 @@
+  displays
+end
+
+def setup_array(files)
```
「setup_array」 → 「配列の準備」だと配列が返るぐらいメソッドぐらいしか分かりません。もう少し具体的にして欲しいです。

ItohYuto replied:
```setup_empty_display_array```に修正しました。
こここのメソッドは表示用配列(display array)の空配列を用意するメソッドのためです。

u1tnk replied:
他でdisplayやarrayは不要と書いてますが、ここは表示するために行列の数値を揃えてるのでこの名前で良いと思いました 👍

ItohYuto replied:
ほかの変数名、メソッド名の変更で```display```を使用しなくなったため、```setup_empty_array```に変更しました。

u1tnk commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,49 @@
+def store_display(files)
+  displays = setup_array(files)
+  max_file_name_length = files.map(&:size).max
+  cnt = 0
```
ループカウンタはruby業務で良く見るのは indexか iですね。短かめのときは略さないのがオススメですが、 i  は慣例でよく使うのでアリです 🙆‍♀️

ItohYuto replied:
```index```に修正しました。
```i```にしなかった理由は```.times```や```.each```のブロックで出てくるとぱっと見で```index```の意味だなと分かりやすい（どういった値が代入されるかがメソッドの時点でわかるため）ですが、定義時点では何に使用されるかが見えないため、```index```にし、配列のインデックス管理用の変数と分かるようにしました。

ループカウンタを```cnt```にするのは癖になっているので、気を付けていきます。

こちらに関して質問なのですが、当初ループカウンタとして使用した```cnt```ですが、今回のような場面では```files```のインデックス管理のために使用されていたため```index```やその省略形の```i```という変数名が適切だと思うのですが、```.times```や```.each```と別でインクリメントしたいけど、その使用先が配列でない場合なども```index```や```i```をループカウンタ名として使用することが多いでしょうか？

u1tnk replied:
そうですね。 `i` だとブロック変数じゃないかと思うのは確かにありますね。スコープが広いほど分かりやすい名前が良いので、`i` ではなくて  `index` にしたのは良いと思います。

> こちらに関して質問なのですが、当初ループカウンタとして使用したcntですが、今回のような場面ではfilesのインデックス管理のために使用されていたためindexやその省略形のiという変数名が適切だと思うのですが、.timesや.eachと別でインクリメントしたいけど、その使用先が配列でない場合などもindexやiをループカウンタ名として使用することが多いでしょうか？

「多い」という話で言うと rubyでは eachや map など ループカウンタを使わないで済むメソッドが多いのでループカウンタ自体が頻出ではありません。
もちろん必要なケースはありますが、大体は each_with_index で eachと一緒に ブロック変数で使うケースが多いですね。
ブロックをまたぐケースになるとあまり頻出ではなく、スコープが広くなるので `xxx_index`など情報量を増やす事が多いと思います。

ItohYuto replied:
メソッドのおかげで単純なループカウンタを使用する機会があまりないということですね。
ありがとうございます。

u1tnk commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,49 @@
+# 表示列
+DISPLAY_COLUMN = 3
+
+def store_display(files)
```
この名前も直訳で 「ディスプレイを保存する」でちょっと理解が難しいです。

ItohYuto replied:
他のメソッドに合わせて接頭語を```setup```にし、```setup_display_array```としました。

ItohYuto replied:
```align_files```に変更しました。

u1tnk commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,49 @@
+DISPLAY_COLUMN = 3
+
+def store_display(files)
+  displays = setup_array(files)
```
displayというのは表示用のデータというイメージでしょうか？
filesを与えて arrayを setupしたら displayの配列ができるというのは理解が難しいので、 arrayの方も考えてもらうとして 語彙を揃えることも考慮して欲しいです。

ItohYuto replied:
メソッド名修正しました。
```store_display```=>```setup_display_array```：表示用配列の準備
```setup_array```=>```setup_empty_display_array```：空の表示用配列の準備
```output_display```=>```display_output```：ディスプレイ出力

u1tnk replied:
私のブログにも書いたのですが、まず ruby界隈では `_array` をほとんど使いません。配列であるかは複数形で示します。Rails の文化の影響が強いからだと思います。
で、「表示用配列」なのですが、用途と型だけであって肝心の中身を示す名称が入っていません。
「xxxを表示するための配列」のxxの部分ですね。
また、用途が色々あるならわかりますが、今回はほぼ表示用なので displayという名前の必要性はあまり無いと思います。

ItohYuto replied:
出力するファイルを整列させているため、```align_files```としました。
戻り値を整列後のファイルとして```aligned_files```にしております。

u1tnk reviewed (commented):
細かいところの前に変数/メソッドの命名を改善して欲しいです。
ちょっと細かいと思われるかもしれませんが、rubyは作者のmatzの座右の銘が「名前重要」という事も一因なのか僕も含めて命名を重視している方が多いという印象です。
その中で仕事をする事を踏まえると厳しめになりますがご了承ください 🙏 

[[初心者向け]変数の命名、基本のキ | u1tnk.log](https://u1tnk.github.io/blog/2022/12/04/naming_variable/) 

こちら僕が以前FBC生向けに書いたブログです。参考になれば 🙏

ItohYuto committed:
timesメソッドの使い方が誤っている個所の修正
配列オブジェクトではなく整数オブジェクトに変更

ItohYuto committed:
不要な後置ifを削除。
rubocop対応。(Use the return of the conditional for variable assignment and comparison.)

ItohYuto committed:
throw,catchを使用せず、ファイルをすべて配列に代入したら、returnでメソッドから抜けるように修正。

ItohYuto committed:
メソッド、変数名の変更。

ItohYuto commented:
修正しました。
再度レビューお願いいたします。
## 実行結果
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/cd840e92-5c70-4057-a068-6ca3d03a36e8)

## rubocop-fjordの実行結果
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/6d7881ca-c298-48f8-812e-e095f4de8e61)

u1tnk commented:
主に変数名についてもう少し検討お願いします 🙏

ItohYuto committed:
メソッド名、変数名の修正。

ItohYuto commented:
定数名変更：```DISPLAY_COLUMN``` => ```MAX_COLUMN```
メソッド名変更：```setup_display_array``` => ```align_files```
                             ```setup_empty_display_array``` => ```setup_empty_array```
                             ```display_output``` => ```output_files```
変数名変更：```display_array``` => ```aligned_files```
                      ```displays``` => ```aligned_files```
                      ```output_displays``` => ```row_strings```

修正しました。
再度レビューお願いいたします。
## 実行結果
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/d479d573-6448-4411-b529-b8f4a7d453a9)

## rubocop-fjordの実行結果
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/17c2888a-c0f7-4f05-a812-3036063df328)

u1tnk commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,46 @@
+
+def align_files(files)
+  aligned_files = setup_empty_array(files)
+  max_file_name_length = files.map(&:size).max
```
このサイズ判定は列ごとに最大にしてください 🙏

ItohYuto replied:
change_width_by_columnメソッドを作成し、「列ごとの最大のファイル名+半角スペース*2」となるように修正しました。
この修正に伴い、2次元配列を```[row][column]```から```[column][row]```に変更しております。

u1tnk commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,46 @@
+  aligned_files = setup_empty_array(files)
+  max_file_name_length = files.map(&:size).max
+  index = 0
+  MAX_COLUMN.times do |column|
```
このままでも良いですが、 `files.each` を全体にして、どこに入れるかを計算するロジックの方がreturnとかも不要だったかなと思ったので一応書いておきます。

ItohYuto replied:
```.each```ではなく```.each_with_index```を使用して修正しました。
```.each_with_index```を使用した理由は、どこに入れるかを計算する際に```files```の何番目か(```index```)が必要になるため、配列の中身(```file```)と何番目か(```index```)が紐づく```.each_with_index```を使用しました。

u1tnk commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,46 @@
+end
+
+def output_files(aligned_files)
+  row_strings = Array.new(aligned_files.size, '')
```
ここはせっかく配置済みの配列があるのでそのまま出力すれば良いですね。
row_stringsは使わずにeachを使ってそのまま表示するようにしましょう。

ItohYuto replied:
おっしゃる通りですね。
そのまま出力するように修正しました。

ItohYuto committed:
ターミナルへの出力時配列の形をそのまま出力するように修正

ItohYuto committed:
each_with_indexを使用して処理を簡略化

ItohYuto committed:
横幅を列ごとに「最大のファイル名+半角スぺース*2」に変更

ItohYuto committed:
rubocop指摘対応

ItohYuto commented:
修正しました。
再度レビューお願いいたします。

## 実行結果
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/fe20a1bb-e08a-4936-a6b1-18ef0a4c50dd)

## rubocop-fjord実行結果
![image](https://github.com/ItohYuto/ruby-practices/assets/88817643/95626773-0833-4e83-9183-3032f66d82c6)

u1tnk reviewed (approved):
問題ありません 🙆‍♀️ 

ちなみに自分で実装するのも学習になるので教えませんでしたが、行列の配列は
https://docs.ruby-lang.org/ja/latest/method/Array/i/transpose.html
を使うと簡単に書けますので参考までに見ておいてください。
